### PR TITLE
[Nicole Joseph] iP

### PR DESCRIPTION
**Pull request by nicolej2122**
###  Ashy ChatBot
_Ashy_ is your friendly NUS cat that also helps with 
 - management of your tasks
 - creation of events 
 - tracking deadlines! 

Start using the _Ashy_ chatbot today! All you need to do is :
1.  Download it from [here](https://github.com/nicolej2122/ip/releases/download/A-Jar/ip.jar)
2. Run the command java `-jar ip.jar` in the same folder as the jar file
3. You can now let _Ashy_  manage your tasks! 😸

 Features: 
- [x] Management of tasks, events and deadlines
- [x] Tasks saved to hard-disk automatically 
- [x]  Search for tasks 

> Download the _Ashy_ ChatBot today for **FREE**!

Here's the main method of the bot!
```
public static void main(String[] args) {
        new Duke().run();
    }
```

### CODE REVIEW

File: src/main/java/duke/Duke.java [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+                System.out.println("You now have " + commands.size() + " tasks in your list");
+            }
+        } catch (DukeException e) {
+            System.out.println(e.getMessage());
+        }
+
+
+    }
+
+    void done(int listNumber) {
```
Comment (GraceWang2322) [inline src/main/java/duke/Duke.java]: maybe use a method name that is easier to understand or more specific??

File: src/main/java/duke/util/Storage.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            }
+
+            if (command.isDone) {
+                text.append("+1\n");
+            } else {
+                text.append("+0\n");
+            }
+        }
+        Path path = Paths.get(directory, "data");
+        boolean exists = Files.exists(path);
```
Comment (GraceWang2322) [inline src/main/java/duke/util/Storage.java]: maybe change boolean variable name to isExists according to the coding standards?

File: src/main/java/duke/util/TaskList.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        }
+
+    }
+
+    /**
+     * Deletes a task from the list 
+     * @param listNumber index of task to be deleted
+     * @param commands tasks in chatbot
+     */
+    public void remove(int listNumber, ArrayList<Task> commands){
```
Comment (GraceWang2322) [inline src/main/java/duke/util/TaskList.java]: maybe have a space before the "{"

File: src/main/java/duke/util/TaskList.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    /**
+     * Finds tasks related to a certain keyword
+     *
+     * @param keyword  keyword to look for in tasks
+     * @param commands all tasks in chatbot
+     */
+    public void find(String keyword, ArrayList<Task> commands) {
+        try {
+            ArrayList<Task> matchingTasks = new ArrayList<>();
+            boolean matchFound = false;
```
Comment (GraceWang2322) [inline src/main/java/duke/util/TaskList.java]: maybe change boolean variable name to isMatchFound

**Review by GraceWang2322 — COMMENTED** [review]
The coding standards are well-followed and the bot is really cutomised!! 👍 just a few nits to review for boolean naming.

File: src/main/java/duke/util/TaskList.java (hunk 4 above)
Comment (nicolej2122) [reply src/main/java/duke/util/TaskList.java]: Noted!

File: src/main/java/duke/Duke.java (hunk 1 above)
Comment (nicolej2122) [reply src/main/java/duke/Duke.java]: the list, done and find methods were named according to their corresponding command in the chatbot

File: src/main/java/duke/task/Event.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    public LocalDateTime time;
+
+    public Event(String description, LocalDateTime time) {
+        super(description);
+        this.time = time;
+    }
+
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (at: " + time.format(DateTimeFormatter.ofPattern("MMM dd yyyy h:mma")) + ")";
```
Comment (EssWhyy) [inline src/main/java/duke/task/Event.java]: This line is slightly long (above the soft limit of 110 characters), perhaps wrap it across 2 lines instead?

File: src/main/java/duke/Duke.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+import duke.exceptions.UnknownInputException;
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Task;
+import duke.task.ToDo;
+import duke.util.DukeCommands;
+import duke.util.Parser;
+import duke.util.Storage;
+import duke.util.TaskList;
+import duke.util.Ui;
```
Comment (EssWhyy) [inline src/main/java/duke/Duke.java]: This is a large group of imports, perhaps consider separating imports of similar subclasses? (e.g. have a line break between duke.exceptions and duke.task and so on.)

File: src/main/java/duke/Duke.java [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+     * @param isDone checks if the task is completed
+     */
+    public String execute(DukeCommands command, String input, int isDone) {
+        DateTimeFormatter formatter1 = DateTimeFormatter.ofPattern("d-MM-yyyy");
+        DateTimeFormatter formatter2 = DateTimeFormatter.ofPattern("d-MM-yyyy HHmm");
+        String response = "";
+        try {
+            boolean taskToAdd = false;
+            int taskIndex;
+            switch (command) {
```
Comment (EssWhyy) [inline src/main/java/duke/Duke.java]: I like the use of ENUMS for your switch block!

**Review by EssWhyy — COMMENTED** [review]
Generally well done, very few coding standard issues. Good use of OOP, javadocs and junit tests throughout the application.

File: src/main/java/duke/Duke.java (hunk 6 above)
Comment (nicolej2122) [reply src/main/java/duke/Duke.java]: I tried that, but it seems to cause a checkstyle violation. I think imports from the same package are all grouped together

File: src/main/java/duke/task/Event.java (hunk 5 above)
Comment (nicolej2122) [reply src/main/java/duke/task/Event.java]: Will do! Thanks!

**Pull request closed without merging**